### PR TITLE
prototype for a 2-step tuning process.

### PR DESCRIPTION
**Pull request by mathiaswagner**
First iterate over full tune space with 2 iterations and keep 5 best candidates.
Do a longer measurement (at least 1e-5s or 3 iterations) with these 5 candidates to find the best launch parameters.

### CODE REVIEW

File: lib/tune.cpp [hunk 1]
```diff
@@ -787,7 +819,8 @@ namespace quda
         param.aux = make_int4(-1, -1, -1, -1);
         tunable.initTuneParam(param);
 
-        while (tuning) {
+        constexpr int candiate_iterations = 2;
```
Comment (maddyscientist) [inline lib/tune.cpp]: probably want to leave this as a variable that can be changed by the user

File: lib/tune.cpp [hunk 2]
```diff
@@ -678,6 +680,35 @@ namespace quda
...
 
+  using tunepair = std::pair<float, TuneParam>;
+  struct tunepaircomp {
+    bool operator()(const tunepair &left, const tunepair &right) const { return (left.first) < (right.first); }
+  };
+
+  class TuneCandidates : public std::priority_queue<tunepair, std::vector<tunepair>, tunepaircomp>
+  {
+  private:
+    static constexpr size_t maxSize = 5;
```
Comment (maddyscientist) [inline lib/tune.cpp]: style comment: prefer `max_size`, I think we should be moving away from camel case, especially for member variables (I'll _tolerate_ it for member functions).

File: lib/tune.cpp [hunk 3]
```diff
@@ -678,6 +680,35 @@ namespace quda
 
   static TimeProfile launchTimer("tuneLaunch");
 
+  using tunepair = std::pair<float, TuneParam>;
```
Comment (maddyscientist) [inline lib/tune.cpp]: Is this new type needed: can't we use the `time` variable already in the `TuneParam`, and just use a queue of `TuneParam`s?

File: lib/tune.cpp (hunk 2 above)
Comment (mathiaswagner) [reply lib/tune.cpp]: good point.

File: lib/tune.cpp [hunk 4]
```diff
@@ -801,7 +834,59 @@ namespace quda
...
+            }
+          }
+          candidatetuning = tunable.advanceTuneParam(param);
+          tunable.launchError() = QUDA_SUCCESS;
+        }
+
+        if (tc.empty()) { errorQuda("Auto-tuning failed for %s with %s at vol=%s", key.name, key.aux, key.volume); }
+
+        constexpr float min_tune_time = 1e-5f;
+        const int tuneiterations = std::max(static_cast<int>(std::ceil(min_tune_time / tc.getBestTime())), 3);
```
Comment (maddyscientist) [inline lib/tune.cpp]: Like the candidate iterations, we should leave the `3` more exposed and not buried in this equation.

Comment (maddyscientist) [reply lib/tune.cpp]: Also, I think the default here should 10.  Given you have the timer cap, there's no reason not to use this.

Comment (mathiaswagner) [reply lib/tune.cpp]: This is to make sure we *at least* do 3 iterations here. The timer cap is to make sure we do more iterations on fast running kernels. And `min_time_tune` should get a sensible default to make sure we don't pick up too much noise / timing inaccuracy.

Comment (maddyscientist) [reply lib/tune.cpp]: Ha, I read it backwards 😆

File: lib/tune.cpp (hunk 1 above)
Comment (weinbe2) [reply lib/tune.cpp]: I'm fine with baking this in at 2, combined with some minimum time requirement.

File: lib/tune.cpp (hunk 2 above)
Comment (weinbe2) [reply lib/tune.cpp]: I think `5` or `10` would be a good default (actually, some preference for `10`, given how much we're thinning the space already). I think this should also be overridable via a `virtual` class function, as well as a global override via an environment variable, where by "global" I mean ignore this value here and all `virtual` values.

The motivation for the "global" override is for development purposes, sometimes it's worthwhile to have a sketchily tuned kernel as opposed to the current alternative of outright disabling tuning.

Comment (mathiaswagner) [reply lib/tune.cpp]: 86b46b64e4871473715e683dfc1ab46d5de8d18c

File: lib/tune.cpp (hunk 3 above)
Comment (mathiaswagner) [reply lib/tune.cpp]: 91c29f86284deea69ca9b6c87fdadcb3525a6cb9

File: lib/tune.cpp (hunk 1 above)
Comment (mathiaswagner) [reply lib/tune.cpp]: we have this as a virtual function now

File: lib/tune.cpp [hunk 5]
```diff
@@ -678,6 +680,54 @@ namespace quda
...
+   */
+  class TuneCandidates : public std::priority_queue<TuneParam, std::vector<TuneParam>, TuneParamComp>
+  {
+  private:
+    const size_t max_size = 5;
+    float besttime = FLT_MAX;
+
+  public:
+    /**
+     * @brief Construct a new Tune Candidates object with specifiwd size
```
Comment (hummingtree) [inline lib/tune.cpp]:
```suggestion
     * @brief Construct a new Tune Candidates object with specified size
```

File: lib/tune.cpp [hunk 6]
```diff
@@ -678,6 +680,54 @@ namespace quda
 
   static TimeProfile launchTimer("tuneLaunch");
 
+  using tunepair = std::pair<float, TuneParam>;
+  struct tunepaircomp {
+    bool operator()(const tunepair &left, const tunepair &right) const { return (left.first) < (right.first); }
+  };
```
Comment (hummingtree) [inline lib/tune.cpp]: This struct is no longer used anymore?

Comment (mathiaswagner) [reply lib/tune.cpp]: Yes, forgot to remove that. Thanks!

File: include/tune_quda.h [hunk 7]
```diff
@@ -243,7 +243,33 @@ namespace quda {
...
+
+    /**
+     * @brief Number of iterations used in the 1st phase of tuning, i.e. finding the candidates for the 2nd phase/
+     *
+     * @return int number of iterations
+     */
+    virtual int candiate_iter() const { return 2; }
+
+    /**
+     * @brief Number of candidates to use for the 2dn tuning phase
```
Comment (maddyscientist) [inline include/tune_quda.h]: typo: 2dn -> 2nd

File: include/tune_quda.h [hunk 8]
```diff
@@ -243,7 +243,33 @@ namespace quda {
     virtual void apply(const qudaStream_t &stream) = 0;
     virtual void preTune() { }
     virtual void postTune() { }
-    virtual int tuningIter() const { return 1; }
+
+    /**
+     * @brief Parameter to control the number of iteration used in the 2nd phase of tuning, i.e. for the candidates.
+     *
+     * @return int minimum number of iterations
```
Comment (maddyscientist) [inline include/tune_quda.h]: The `return` statement shouldn't include the type being returned (same applies for the next few functions in this file).

File: include/tune_quda.h [hunk 9]
```diff
@@ -243,7 +243,33 @@ namespace quda {
...
+     */
+    virtual int min_tune_iter() const { return 3; }
```
Comment (maddyscientist) [inline include/tune_quda.h]: I think the order of these declarations should be reversed, such that the parameters for the first stage of tuning should come first, and the candidate (2nd) tuning should come second.

**Review by maddyscientist — COMMENTED** [review]
Great addition.  Beyond the quick-to-fix comments I've left, this PR is good to go.  I'll approve once those fixes have been applied.

**Review by maddyscientist — APPROVED** [review]

**Comment by mathiaswagner** [comment]
Before merging that I need to run some more test with Chroma + NVSHMEM.

**Comment by mathiaswagner** [comment]
Did some more tests and things are looking good now.

**Review by weinbe2 — APPROVED** [review]
This has passed all my various correctness and perf tests, so this is good to go. Great contribution @mathiaswagner !

**Pull request merged**